src/thread_event.c

```diff
@@ -0,0 +1,277 @@
+#include "jemalloc/internal/jemalloc_internal_includes.h"
+
+#include "jemalloc/internal/thread_event.h"
+
+/*
+ * If we don't set an upper limit for event triggering which is significantly
+ * smaller than UINT64_MAX, then if all events have very long wait times, when
+ * the threshold for the next event rollovers uint64_t, it can stay there for a
+ * very long time, during which the malloc fast path check will always fail and
+ * the execution will always fall back to slow path.
+ */
+#define THREAD_EVENT_MAX_INTERVAL ((uint64_t)1048576)
+
+/* Event handle function signatures. */
+#define W(event, condition)						\
+static void thread_##event##_event_handle(tsd_t *tsd);
+
+ITERATE_OVER_ALL_WAITS
+#undef W
+
+static void
+thread_event_wait_assert(uint64_t wait) {
+	assert(wait == 0U ||
+	    (wait >= THREAD_EVENT_MIN_WAIT && wait <= THREAD_EVENT_MAX_WAIT));
+}
+
+static uint64_t
+thread_allocated_next_event_compute(tsd_t *tsd, bool *no_event_on_p) {
+	uint64_t wait = THREAD_EVENT_MAX_WAIT;
+	*no_event_on_p = true;
+
+#define W(event, condition)						\
+	if (condition) {						\
+		*no_event_on_p = false;					\
+		uint64_t event_wait =					\
+		    event##_event_wait_get(tsd);			\
+		thread_event_wait_assert(event_wait);			\
+		if (event_wait > 0 && event_wait < wait) {		\
+			wait = event_wait;				\
+		}							\
+	}
+
+	ITERATE_OVER_ALL_WAITS
+#undef W
+
+	thread_event_wait_assert(wait);
+	return wait;
+}
+
+void
+thread_event_assert_invariants_debug(tsd_t *tsd) {
+	uint64_t thread_allocated = thread_allocated_get(tsd);
+	uint64_t last_event = thread_allocated_last_event_get(tsd);
+	uint64_t next_event = thread_allocated_next_event_get(tsd);
+	uint64_t next_event_fast = thread_allocated_next_event_fast_get(tsd);
+
+	assert(last_event != next_event);
+	if (next_event <= THREAD_ALLOCATED_NEXT_EVENT_FAST_MAX) {
+		assert(next_event_fast == next_event);
+	} else {
+		assert(next_event_fast == 0);
+	}
+
+	/* The subtraction is intentionally susceptible to underflow. */
+	uint64_t interval = next_event - last_event;
+
+	/* The subtraction is intentionally susceptible to underflow. */
+	assert(thread_allocated - last_event < interval);
+
+	bool no_event_on;
+	uint64_t min_wait = thread_allocated_next_event_compute(tsd,
+	    &no_event_on);
+
+	if (no_event_on) {
+		/*
+		 * next_event should have been pushed up only except when the
+		 * TSD is just initialized.  The last_event == 0 guard below is
+		 * stronger than needed, but having an exactly accurate guard
+		 * is more complicated to implement.
+		 */
+		assert(last_event == 0 ||
+		    next_event == THREAD_ALLOCATED_NEXT_EVENT_FAST_MAX ||
+		    next_event == THREAD_ALLOCATED_NEXT_EVENT_FAST_MAX +
+		    THREAD_EVENT_MIN_WAIT);
+	} else {
+		assert(interval == min_wait || (interval < min_wait &&
+		    interval == THREAD_EVENT_MAX_INTERVAL));
+	}
+}
+
+static void
+thread_event_adjust_thresholds_helper(tsd_t *tsd, uint64_t last_event,
+    uint64_t wait, bool no_event_on) {
```

Comment:
The no event on case should be contained within this function. Move the logic in event_trigger to here, and handle that case in a separate branch, e.g. 

    If (no event on) {
        wait = 
    } else {
        wait = ...
    }

I would even say, since no event is going to be a rare case, let’s not optimize for it. Let the max interval take care that case is fine, i.e. not keeping the trick for no event on.

Reply:
No event on may not be a rare case? If `opt_prof` is off, then there's no event on.

Reply:
Since we’ll add gc ticker as well, it should be rare. I mean even in that case, the penalty is minimal, only trigger once medium path and event update every max interval?

Reply:
Yes, we can do that for no event on case, but what I had in mind is, it's serving the same purpose as that for setting the threshold to `THREAD_ALLOCATED_NEXT_EVENT_FAST_MAX`. If we want to have the max interval for all cases, then should we get rid of the `THREAD_ALLOCATED_NEXT_EVENT_FAST_MAX` trick?

Reply:
Yeah I’d vote for removing the trick of setting to event_fast_max. The benefit is small enough that we probably shouldn’t bite the complexity.

Reply:
Makes sense. Now `thread_event_adjust_thresholds_helper()` can be independent of whether any event is on or not, so I further cleaned up the logic.